File: bindings/dotnet/perf/single/Zlink.BindingBench/src/PerfGateway.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading;
using Zlink;
using static PerfRunner;

internal static class PerfGateway
{
    internal static int RunGateway(string transport, int size)
    {
        int warmupCount = ResolveSingleWarmupCount("GATEWAY");
        int settleMs = SingleSettleTimeMs;
        int durationSeconds = ParseEnv("PERF_SINGLE_DURATION_SECONDS", 5);
        int recvTimeoutMs = ParseEnvNonNegative("PERF_SINGLE_RCVTIMEO_MS", 200);
        int latCount = ResolveSingleLatencyCount("GATEWAY");

        using var ctx = new Context();
        ApplySingleContextOptions(ctx);
        Registry? registry = null;
        Discovery? discovery = null;
        Receiver? receiver = null;
        Gateway? gateway = null;

        try
        {
            string suffix =
                $"{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}-{Guid.NewGuid()}";
            string regPub = EndpointFor("tcp", $"gw-pub-{suffix}");
            string regRouter = EndpointFor("tcp", $"gw-router-{suffix}");
            string service = "svc";

            registry = new Registry(ctx);
            registry.SetHeartbeat(5000, 60000);
            registry.SetEndpoints(regPub, regRouter);
            registry.Start();

            discovery = new Discovery(ctx, DiscoveryServiceType.Gateway);
            discovery.ConnectRegistry(regRouter);
            receiver = new Receiver(ctx);
            string providerEp = EndpointFor(transport, "gateway-provider");
            ConfigureReceiverTlsServerIfNeeded(receiver, transport);
            receiver.Bind(providerEp);
            receiver.ConnectRegistry(regRouter);
            receiver.Register(service, providerEp, 1);

            gateway = new Gateway(ctx, discovery);
            ConfigureGatewayTlsClientIfNeeded(gateway, transport);
            Gateway.PreparedService preparedService =
                gateway.PrepareService(service);
            int readyTimeoutMs = ResolveGatewayReadyTimeoutMs();

            if (!WaitUntil(() => discovery.ReceiverCount(service) > 0,
                    readyTimeoutMs, 1)
                || !WaitUntil(() => gateway.ConnectionCount(service) > 0,
                    readyTimeoutMs, 1))
            {
                return 2;
            }

            int payloadSize = Math.Max(size, sizeof(long));
            var payload = new byte[payloadSize];
            Array.Fill(payload, (byte)'a');

            if (!RunPhase(gateway, receiver, preparedService, payload, payloadSize,
                    warmupCount, 0, recvTimeoutMs, 0, out long warmupReceived,
                    out _)
                || warmupReceived < warmupCount)
            {
                return 2;
            }

            Thread.Sleep(settleMs);

            if (!RunPhase(gateway, receiver, preparedService, payload, payloadSize, 0,
                    durationSeconds, recvTimeoutMs, latCount, out long received,
                    out var latencySamples))
            {
                return 2;
            }

            double throughput = received / (double)Math.Max(durationSeconds, 1);
            var latency = ComputeLatencyStats(latencySamples);
            PrintResult("GATEWAY", transport, size, throughput, latency.mean,
                latency.p95, latency.p99);
            return 0;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"single_gateway_error:{ex.Message}");
            return 2;
        }
        finally
        {
            TryDisposeAllQuietly(gateway, receiver, discovery, registry);
        }
    }

    private static bool RunPhase(Gateway gateway, Receiver receiver,
        Gateway.PreparedService service, byte[] payload, int payloadSize,
        int warmupCount, int durationSeconds, int recvTimeoutMs, int latencyCap,
        out long receivedOut, out List<double> latencySamples)
    {
        bool active = durationSeconds > 0;
        long deadlineTicks = active
            ? DeadlineTicksFromSeconds(durationSeconds)
            : 0;
        long drainTicks = Math.Max(1,
            (long)Math.Ceiling(recvTimeoutMs * Stopwatch.Frequency / 1000.0));

        long received = 0;
        int senderDone = 0;
        Exception? recvError = null;
        var samples = new List<double>(Math.Max(0, latencyCap));
        long sampleSeen = 0;
        uint rng = 0xA341316Cu;

        var recvThread = new Thread(() =>
        {
            long lastRecvTicks = Stopwatch.GetTimestamp();
            var recvBuffer = new byte[payloadSize];

            void AccountMessage(int bytesRead)
            {
                if (bytesRead != payloadSize)
                    return;

                Interlocked.Increment(ref received);
                if (!active)
                    return;

                long nowUs = TimestampUs();
                long sentUs = DecodeHeader(recvBuffer.AsSpan(0, bytesRead));
                double latencyUs = Math.Max(0L, nowUs - sentUs);
                ReservoirSample(samples, latencyUs, ref sampleSeen, latencyCap,
                    ref rng);
            }

            try
            {
                while (true)
                {
                    bool done = Volatile.Read(ref senderDone) != 0;
                    try
                    {
                        int bytesRead = receiver.ReceiveSinglePayload(recvBuffer,
                            done ? ReceiveFlags.DontWait : ReceiveFlags.None);
                        lastRecvTicks = Stopwatch.GetTimestamp();
                        AccountMessage(bytesRead);

                        while (true)
                        {
                            bytesRead = receiver.ReceiveSinglePayload(recvBuffer,
                                ReceiveFlags.DontWait);
                            lastRecvTicks = Stopwatch.GetTimestamp();
                            AccountMessage(bytesRead);
                        }
                    }
                    catch (ZlinkException ex) when (IsInterrupted(ex.Errno))
                    {
                        continue;
                    }
                    catch (ZlinkException ex) when (IsWouldBlock(ex.Errno))
                    {
                        if (done && Stopwatch.GetTimestamp() - lastRecvTicks
                            >= drainTicks)
                        {
                            break;
                        }
                        Thread.Yield();
                    }
                }
            }
            catch (Exception ex)
            {
                recvError = ex;
            }
        });
        recvThread.IsBackground = true;
        recvThread.Start();

        bool sendFailed = false;
        if (active)
        {
            while (Stopwatch.GetTimestamp() < deadlineTicks)
            {
                StampHeader(payload.AsSpan(0, sizeof(long)), TimestampUs());
                try
                {
                    gateway.Send(service, payload.AsSpan(), SendFlags.None);
                }
                catch
                {
                    sendFailed = true;
                    break;
                }
            }
        }
        else
        {
            for (int i = 0; i < warmupCount; i++)
            {
                StampHeader(payload.AsSpan(0, sizeof(long)), TimestampUs());
                try
                {
                    gateway.Send(service, payload.AsSpan(), SendFlags.None);
                }
                catch
                {
                    sendFailed = true;
                    break;
                }
            }
        }

        Volatile.Write(ref senderDone, 1);
        recvThread.Join();

        latencySamples = samples;
        receivedOut = received;
        if (sendFailed || recvError != null)
            return false;

        if (!active)
            return received >= warmupCount;

        return received > 0 && latencySamples.Count > 0;
    }

    private static bool IsInterrupted(int errno)
    {
        ErrorCode code = ZlinkException.MapErrorCode(errno);
        return code == ErrorCode.EIntr || errno == 4;
    }

    private static bool IsWouldBlock(int errno)
    {
        ErrorCode code = ZlinkException.MapErrorCode(errno);
        return code == ErrorCode.EAgain || errno == 11;
    }

}

File: bindings/java/perf/single/Zlink.PerfBench/src/main/java/dev/kairoscode/zlink/integration/bench/src/PerfGateway.java
/* SPDX-License-Identifier: MPL-2.0 */

package dev.kairoscode.zlink.integration.bench.src;

import dev.kairoscode.zlink.Context;
import dev.kairoscode.zlink.ReceiveFlag;
import dev.kairoscode.zlink.SendFlag;
import dev.kairoscode.zlink.ServiceType;
import dev.kairoscode.zlink.ZlinkException;
import dev.kairoscode.zlink.integration.bench.common.PerfCommon;
import dev.kairoscode.zlink.integration.bench.common.PerfSingleMetricHeader;
import dev.kairoscode.zlink.integration.bench.common.PerfTls;
import dev.kairoscode.zlink.options.SocketOptions;
import dev.kairoscode.zlink.service.discovery.Discovery;
import dev.kairoscode.zlink.service.gateway.Gateway;
import dev.kairoscode.zlink.service.receiver.Receiver;
import dev.kairoscode.zlink.service.registry.Registry;
import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;
import java.util.concurrent.atomic.LongAdder;

/**
 * GATEWAY benchmark using high-level Gateway/Receiver APIs.
 */
public final class PerfGateway {
    private static final String PATTERN = "GATEWAY";

    private static final int ERRNO_EINTR = 4;
    private static final int ERRNO_EAGAIN = 11;
    private static final int ERRNO_EWOULDBLOCK_WIN = 10035;

    private PerfGateway() {
    }

    public static int run(String transport, int msgSize) {
        String tr = transport == null ? "" : transport.toLowerCase();
        if ("inproc".equals(tr) || "ipc".equals(tr)) {
            PerfCommon.printUnsupported(PATTERN, tr, msgSize,
                "gateway does not support inproc/ipc");
            return 0;
        }

        String unique = Long.toHexString(System.nanoTime());
        String regPub = "inproc://java-perf-gw-pub-" + unique;
        String regRouter = "inproc://java-perf-gw-router-" + unique;
        String serviceName = "svc";

        try (Context context = new Context();
             Registry registry = new Registry(context);
             Discovery discovery = new Discovery(context, ServiceType.GATEWAY);
             Receiver receiver = new Receiver(context);
             Gateway gateway = new Gateway(context, discovery)) {
            PerfCommon.applySingleContextOptions(context);

            registry.setEndpoints(regPub, regRouter);
            registry.setHeartbeat(5000, 60000);
            registry.start();

            connectRegistryWithRetry(() -> discovery.connectRegistry(regRouter));
            discovery.subscribe(serviceName);

            applyServiceSocketOptions(gateway, receiver);

            String providerEndpoint = PerfCommon.endpointFor(tr,
                "gateway-provider");
            PerfTls.configureReceiverTlsServerIfNeeded(receiver, tr);
            receiver.bind(providerEndpoint);
            connectRegistryWithRetry(() -> receiver.connectRegistry(regRouter));
            receiver.register(serviceName, providerEndpoint, 1);

            PerfTls.configureGatewayTlsClientIfNeeded(gateway, tr);
            int readyTimeoutMs = PerfCommon.resolveGatewayReadyTimeoutMs();

            if (!PerfCommon.waitUntil(
                () -> receiver.registerResult(serviceName).status() == 0,
                readyTimeoutMs, 1)) {
                return 2;
            }

            if (!PerfCommon.waitUntil(() -> discovery.receiverCount(serviceName) > 0,
                readyTimeoutMs, 1)) {
                return 2;
            }

            if (!PerfCommon.waitUntil(() -> gateway.connectionCount(serviceName) > 0,
                readyTimeoutMs, 1)) {
                return 2;
            }

            int payloadSize = Math.max(msgSize, PerfSingleMetricHeader.HEADER_SIZE);
            int warmupCount = PerfCommon.resolveWarmupCount(PATTERN, msgSize);
            int durationSeconds = PerfCommon.resolveDurationSeconds();
            int latencyCap = PerfCommon.resolveLatencySampleCap();
            int recvTimeoutMs = PerfCommon.resolveRecvTimeoutMs();
            int runId = PerfCommon.randomRunId();

            try (Arena arena = Arena.ofConfined()) {
                MemorySegment payload = arena.allocate(payloadSize, 8);
                payload.fill((byte) 'a');
                long[] seq = new long[] {1L};

                PhaseResult warmup = runPhase(gateway, receiver, serviceName,
                    payload, payloadSize, msgSize, runId, seq,
                    PerfSingleMetricHeader.PHASE_WARMUP, warmupCount, 0, 0,
                    recvTimeoutMs);
                if (!warmup.ok || warmup.received < warmupCount) {
                    return 2;
                }

                Thread.sleep(PerfCommon.resolveSettleMs());

                PhaseResult active = runPhase(gateway, receiver, serviceName,
                    payload, payloadSize, msgSize, runId, seq,
                    PerfSingleMetricHeader.PHASE_ACTIVE, 0, durationSeconds,
                    latencyCap, recvTimeoutMs);
                if (!active.ok || active.received <= 0) {
                    return 2;
                }

                double throughput =
                    active.received / (double) Math.max(durationSeconds, 1);
                PerfCommon.Stats stats = active.latency.snapshot();
                PerfCommon.printResult(PATTERN, tr, msgSize, throughput,
                    stats.meanUs(), stats.p95Us(), stats.p99Us());
                return 0;
            }
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
            return 2;
        } catch (RuntimeException ex) {
            String message = ex.getMessage() == null
                ? ex.getClass().getSimpleName()
                : ex.getMessage();
            System.err.println("single_gateway_error:" + message);
            return 2;
        }
    }

    private static PhaseResult runPhase(Gateway gateway, Receiver receiver,
                                        String serviceName,
                                        MemorySegment payload, int payloadSize,
                                        int msgSize, int runId, long[] seq,
                                        int phase, int warmupCount,
                                        int durationSeconds, int latencyCap,
                                        int recvTimeoutMs) {
        final boolean active = durationSeconds > 0;
        final long deadlineNs = active
            ? System.nanoTime()
                + (long) Math.max(durationSeconds, 1) * 1_000_000_000L
            : 0L;
        final long drainDeadlineNs =
            (long) Math.max(recvTimeoutMs, 1) * 1_000_000L;
        final LongAdder received = new LongAdder();
        final AtomicBoolean senderDone = new AtomicBoolean(false);
        final AtomicReference<RuntimeException> recvError =
            new AtomicReference<>();
        final PerfCommon.LatencyReservoir latency =
            new PerfCommon.LatencyReservoir(Math.max(latencyCap, 1));

        Thread receiverThread = new Thread(() -> {
            long lastRecvNs = System.nanoTime();
            PerfSingleMetricHeader.Header header =
                new PerfSingleMetricHeader.Header();

            try (Arena recvArena = Arena.ofConfined()) {
                MemorySegment recvBuffer = recvArena.allocate(payloadSize, 8);
                while (true) {
                    boolean done = senderDone.get();
                    try {
                        int bytesRead = receiver.recvSinglePayload(recvBuffer,
                            done ? ReceiveFlag.DONTWAIT : ReceiveFlag.NONE);
                        lastRecvNs = System.nanoTime();
                        accountMessage(recvBuffer, bytesRead, payloadSize,
                            active, runId, phase, header, received, latency);

                        while (true) {
                            bytesRead = receiver.recvSinglePayload(recvBuffer,
                                ReceiveFlag.DONTWAIT);
                            lastRecvNs = System.nanoTime();
                            accountMessage(recvBuffer, bytesRead, payloadSize,
                                active, runId, phase, header, received, latency);
                        }
                    } catch (ZlinkException ex) {
                        if (isInterrupted(ex.errno())) {
                            continue;
                        }
                        if (isWouldBlock(ex.errno())) {
                            if (done
                                && System.nanoTime() - lastRecvNs >= drainDeadlineNs) {
                                break;
                            }
                            Thread.yield();
                            continue;
                        }
                        throw ex;
                    }
                }
            } catch (RuntimeException ex) {
                recvError.set(ex);
            }
        }, "zlink-java-perf-gateway-recv");
        receiverThread.setDaemon(true);
        receiverThread.start();

        boolean ok = true;
        if (active) {
            while (System.nanoTime() < deadlineNs) {
                if (!PerfSingleMetricHeader.stampPayload(payload, payloadSize,
                    runId, phase, msgSize, seq[0]++,
                    PerfSingleMetricHeader.nowUs())) {
                    ok = false;
                    break;
                }
                try {
                    gateway.sendTo(serviceName, payload, SendFlag.NONE);
                } catch (RuntimeException ex) {
                    ok = false;
                    break;
                }
            }
        } else {
            for (int i = 0; i < warmupCount; i++) {
                if (!PerfSingleMetricHeader.stampPayload(payload, payloadSize,
                    runId, phase, msgSize, seq[0]++,
                    PerfSingleMetricHeader.nowUs())) {
                    ok = false;
                    break;
                }
                try {
                    gateway.sendTo(serviceName, payload, SendFlag.NONE);
                } catch (RuntimeException ex) {
                    ok = false;
                    break;
                }
            }
        }

        senderDone.set(true);
        try {
            receiverThread.join();
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
            return new PhaseResult(false, received.sum(), latency);
        }

        return new PhaseResult(ok && recvError.get() == null, received.sum(),
            latency);
    }

    private static void accountMessage(MemorySegment recvBuffer, int bytesRead,
                                       int payloadSize, boolean active,
                                       int runId, int phase,
                                       PerfSingleMetricHeader.Header header,
                                       LongAdder received,
                                       PerfCommon.LatencyReservoir latency) {
        if (bytesRead != payloadSize) {
            return;
        }
        if (!PerfSingleMetricHeader.decodePayloadHeader(recvBuffer, payloadSize,
            header)) {
            return;
        }
        if (header.runId != runId || header.phase != phase) {
            return;
        }

        received.increment();
        if (!active) {
            return;
        }
        long nowUs = PerfSingleMetricHeader.nowUs();
        latency.add(Math.max(0L, nowUs - header.sentTsUs));
    }

    private static void applyServiceSocketOptions(Gateway gateway,
                                                  Receiver receiver) {
        int sndHwm = resolveSingleHwmValue("PERF_SINGLE_SNDHWM");
        int rcvHwm = resolveSingleHwmValue("PERF_SINGLE_RCVHWM");
        int sndTimeo = PerfCommon.parseEnvNonNegative(
            "PERF_SINGLE_SNDTIMEO_MS", 200);
        int recvTimeo = PerfCommon.parseEnvNonNegative(
            "PERF_SINGLE_RCVTIMEO_MS", 200);

        gateway.setOption(SocketOptions.SNDHWM, sndHwm);
        gateway.setOption(SocketOptions.RCVHWM, rcvHwm);
        gateway.setOption(SocketOptions.SNDTIMEO, sndTimeo);
        gateway.setOption(SocketOptions.RCVTIMEO, recvTimeo);

        receiver.setOption(SocketOptions.SNDHWM, sndHwm);
        receiver.setOption(SocketOptions.RCVHWM, rcvHwm);
        receiver.setOption(SocketOptions.SNDTIMEO, sndTimeo);
        receiver.setOption(SocketOptions.RCVTIMEO, recvTimeo);
    }

    private static int resolveSingleHwmValue(String specificName) {
        int hwm = PerfCommon.parseEnv("PERF_SINGLE_HWM", 1000);
        int specific = PerfCommon.parseEnv(specificName, 0);
        return specific > 0 ? specific : hwm;
    }

    private static boolean isWouldBlock(int errno) {
        return errno == ERRNO_EAGAIN || errno == ERRNO_EWOULDBLOCK_WIN;
    }

    private static boolean isInterrupted(int errno) {
        return errno == ERRNO_EINTR;
    }

    private static void connectRegistryWithRetry(Runnable connect) {
        RuntimeException last = null;
        long deadlineNs = System.nanoTime() + 5_000_000_000L;
        while (System.nanoTime() < deadlineNs) {
            try {
                connect.run();
                return;
            } catch (ZlinkException ex) {
                if (!isInterrupted(ex.errno()) && !isWouldBlock(ex.errno())) {
                    throw ex;
                }
                last = ex;
            }
            try {
                Thread.sleep(20);
            } catch (InterruptedException ex) {
                Thread.currentThread().interrupt();
                throw new IllegalStateException("registry connect interrupted", ex);
            }
        }

        if (last != null) {
            throw last;
        }
        throw new IllegalStateException("registry connect timeout");
    }

    private record PhaseResult(boolean ok, long received,
                               PerfCommon.LatencyReservoir latency) {
    }
}
